File: fordfulkerson.py
#!/usr/bin/env python3
from collections import deque
import sys

def bfs(rgraph, s, e, parent):
    n = len(rgraph)
    visited = [False] * n
    q = deque()
    q.append(s)
    visited[s] = True
    parent[s] = -1

    while q:
        curr = q.popleft()
        for i in range(n):
            if rgraph[curr][i] > 0 and not visited[i]:
                visited[i] = True
                parent[i] = curr
                q.append(i)
                if i == e:
                    return True
    return False

def ford_fulkerson(g, s, e):
    n = len(g)
    max_flow = 0
    parent = [-1] * n
    # build residual graph as n x n matrix
    rgraph = [[0] * n for _ in range(n)]
    for u in range(n):
        for (v, cap) in g[u]:
            rgraph[u][v] = cap

    valid_paths = []

    while bfs(rgraph, s, e, parent):
        # find bottleneck
        node = e
        bottleneck = float('inf')
        while parent[node] != -1:
            par = parent[node]
            bottleneck = min(bottleneck, rgraph[par][node])
            node = par

        # update residual graph along path
        node = e
        while parent[node] != -1:
            par = parent[node]
            rgraph[par][node] -= bottleneck
            rgraph[node][par] += bottleneck
            node = par

        max_flow += bottleneck

        # build path (forward order)
        rev_path = []
        node = e
        while parent[node] != -1:
            rev_path.append(node)
            node = parent[node]
        rev_path.append(node)
        rev_path.reverse()
        valid_paths.append(rev_path)

        # reset parent for next BFS (bfs will set it again)
        parent = [-1] * n

    # print results to stdout
    print(int(max_flow))
    for path in valid_paths:
        print(" ".join(str(x) for x in path))

def main():
    # read from filename arg if provided, otherwise from stdin
    if len(sys.argv) > 1:
        with open(sys.argv[1], "r") as f:
            data = f.read().strip().split()
    else:
        data = sys.stdin.read().strip().split()

    if not data:
        return

    it = iter(data)
    try:
        N = int(next(it))
    except StopIteration:
        return
    E = int(next(it))
    g = [[] for _ in range(N)]
    for _ in range(E):
        src = int(next(it))
        dst = int(next(it))
        capacity = int(next(it))
        g[src].append((dst, capacity))
    s = int(next(it))
    e = int(next(it))
    ford_fulkerson(g, s, e)

if __name__ == "__main__":
    main()

# python3 fordfulkerson.py input.txt

File: fordfulkerson.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <queue>
#include <climits>

using namespace std;

bool BFS(const vector<vector<int>> &rGraph, int s, int e, vector<int> &parent, int vToken, vector<int> &visited)
{
    int N = rGraph.size();
    queue<int> q;
    q.push(s);
    visited[s] = vToken;
    parent[s] = -1;

    while (!q.empty())
    {
        int curr = q.front();
        q.pop();
        for (int i = 0; i < N; i++)
        {
            if (rGraph[curr][i] > 0 && visited[i] != vToken)
            {
                visited[i] = vToken;
                parent[i] = curr;
                q.push(i);
                if (i == e)
                {
                    return true;
                }
            }
        }
    }
    return false;
}

void fordFulkerson(const vector<vector<pair<int, int>>> &g, int s, int e)
{
    int N = g.size();
    int maxFlow = 0;
    vector<int> parent(N);
    vector<vector<int>> rGraph(N, vector<int>(N, 0));
    vector<vector<int>> validPaths;

    // build residual graph from adjacency list
    for (int i = 0; i < N; i++)
    {
        for (auto &nei : g[i])
        {
            int dst = nei.first;
            int cap = nei.second;
            rGraph[i][dst] = cap;
        }
    }

    vector<int> visited(N, 0);
    int vToken = 1;
    while (BFS(rGraph, s, e, parent, vToken, visited))
    {
        int bottleKneck = INT_MAX;

        // find min flow in the path using parent
        int node = e;
        while (parent[node] != -1)
        {
            int par = parent[node];
            bottleKneck = min(bottleKneck, rGraph[par][node]);
            node = par;
        }

        // subtract bottlekneck from the path in residual
        node = e;
        while (parent[node] != -1)
        {
            int par = parent[node];
            rGraph[par][node] -= bottleKneck;
            rGraph[node][par] += bottleKneck;
            node = par;
        }
        maxFlow += bottleKneck;
        vToken++;

        // build valid path
        vector<int> revPath;
        node = e;
        while (parent[node] != -1)
        {
            int par = parent[node];
            revPath.push_back(node);
            node = par;
        }
        revPath.push_back(node);
        reverse(revPath.begin(), revPath.end());
        validPaths.push_back(revPath);
    }

    cout << maxFlow << "\n";
    // print all paths

    for (auto &vec : validPaths)
    {
        for (auto &node : vec)
        {
            cout << node << " ";
        }
        cout << "\n";
    }
}

// This algorithm is to find max flow in a network
int main()
{

    int N, E;
    cin >> N >> E;

    vector<vector<pair<int, int>>> g(N);
    for (int i = 0; i < E; i++)
    {
        int src, dst, capacity;
        cin >> src >> dst >> capacity;
        // src--,dst--;
        g[src].emplace_back(dst, capacity);
    }
    int s, e;
    cin >> s >> e;
    fordFulkerson(g, s, e);

    return 0;
}

// g++ -std=c++17 -O2 -Wall -Wextra fordfulkerson.cpp -o fordfulkerson
// Get-Content input.txt -Raw | .\fordfulkerson.exe